02-peer-protocol.md

```diff
@@ -445,6 +446,177 @@ to broadcast the commitment transaction to get his funds back and open a new
+
+1. type: 76 (`splice_ack`)
+2. data:
+   * [`chain_hash`:`chain_hash`]
+   * [`channel_id`:`channel_id`]
+   * [`point`:`funding_pubkey`]
+
+#### Requirements
+
+The sender of `splice`:
+- MUST have successfully initiated quiescence.
+- MUST NOT send `splice` before sending and receiving `funding_locked`.
+- MUST NOT send `splice` while a splice is being negotiated.
+- If any splice is in progress:
+  - MUST NOT send a splice message with `funding_feerate_perkw` which is less than 1.25 the previous `funding_feerate_perkw` (rounded down).
+
+The receiver of `splice`:
+- SHOULD fail the connection if there is an ongoing splice, and the `funding_feerate_perkw` is not at least 1.25 the previous `funding_feerate_perkw` (rounded down).
+- MUST respond with `splice_ack` containing its own `funding_pubkey`.
+- MUST begin splice negotiation.
+
+The receiver of `splice_ack`:
+- MUST begin splice negotiation.
+
+
+#### Rationale
+
+There is no harm in agreeing to a splice with a high feerate
+(presumably the recipient will not contribute to the splice which they
+consider to be overpaying).
+
+## Splice Negotiation
+
+The splice negotiation is very similar to the `init_rbf` negotiation:
+both sides alternate sending `tx_add_input` and `tx_add_output` until
+they both send consecutive `tx_complete`.
+
+### Requirements
+
+(Note that many of these messages have their own, additional
+requirements detailed elsewhere).
+
+The initiator:
+- MUST `tx_add_input` an input which spends the current funding transaction output.
+- MUST `tx_add_output` a zero-value output which pays to the two funding keys using the higher of the two `generation` fields.
+- MUST pay for the common fields.
+
+Upon receipt of consecutive `tx_complete`s, each node:
+- MUST fail negotiation if there is not exactly one input spending the current funding transaction.
+- MUST fail negotiation if there is not exactly one output with zero value paying to the two funding keys (a.k.a. the new channel funding output)
+- MUST calculate the channel capacity for each side:
+  - Start with the previous balance
+  - Add that side's new inputs (excluding the one spending the current funding transaction)
+  - Subtracting each sides new outputs (except the zero-value one paying to the funding keys)
+  - Subtract the total fee that side is paying for the splice transaction.
+- MUST replace the zero-value funding output amount with the total channel capacity.
+- MUST calculate the channel balance for each side:
+  - Subtract any outstanding HTLCs offered by that side.
+- if either side has added an output other than the new channel funding output:
+  - MUST fail the negotiation if the balance for that side is less than 1% of the total channel capacity.
+- SHOULD NOT fail if the splice transaction is nonstandard.
+- MUST increment the commitment number and send `commitment_signed`, including the signatures for the splice transaction.
+
+- Upon receipt of `revoke_and_ack` for the previous commitment:
```

Comment:
I know this is an old draft, but I think the `revoke_and_ack` should be the last step:

```
--cs----->
<-cs------
--tx_sig-->
<-tx_sig---
--rev----->
<-rev-----
```
Otherwise, it's possible that one side has revoked their prior commitment and now has a commit_sig for a tx they can't broadcast as they don't have tx_sig yet.

Reply:
Thanks for pointing this out.

The current flow is:
```
--cs----->
<--rev---
<--cs----
---rev--->
```

Then later we do the `tx_sig` stuff.

The rev messages are implied here but are specified elsewhere in the spec. Here's the current process showing how the rev messages are handled: https://github.com/ddustin/lightning/blob/5b655039d0f87acf2b2fec2cac11b999ab2790f3/channeld/channeld.c#L2500

Could be worth adding the rev's to the spec here for additional clarity 🤷

Reply:
B _gets_ a signature from A (and has 2 non-revocable commitments then), then B revokes the previous one. one of the parties either has 1 or 2 valid commitments, not 0

Reply:
If I'm understanding this correctly, B's commit_sig that A sends is spending from the intermediate splice transaction. It is not usable until B has the tx_sig from A to broadcast the splice transaction.

Reply:
oh you're right, the flow in your [first comment](https://github.com/lightning/bolts/pull/863#discussion_r910126737) is the way to go then, or to make it a bit faster:
```
--cs------>
<-cs-------    <-- 2 messages at once
<-tx_sig---
--tx_sig-->    <-- 2 messages at once
--rev----->
<-rev------
```
but this can make things more complicated

Reply:
Think my comment is wrong here since the `commit_sig` actually includes a signature spending from the original funding outpoint and a signature (in the tlv) spending from the new funding outpoint

Reply:
if that is the case i think this line has to be made clearer

```
- MUST increment the commitment number and send `commitment_signed`, including the signatures for the splice transaction.
```

are these "signatures for the splice transaction" actually signatures for the commitment transaction that spends the splice transaction?

Reply:
They're incrementing the commitment tx signatures (not signing the splice tx itself -- there's no count for those that I'm aware of).

I believe the commitment number is shared across both original funding and splice funding.

Reply:
> if that is the case i think this line has to be made clearer
> 
> ```
> - MUST increment the commitment number and send `commitment_signed`, including the signatures for the splice transaction.
> ```
> 
> are these "signatures for the splice transaction" actually signatures for the commitment transaction that spends the splice transaction?

These are the signatures for the commitment transaction child of the splice -- I'll update it to make that more clear 👍